### PR TITLE
Adding VPN options for IKE version and IKE split connections.

### LINKED ISSUE #3138: StrongSwan with several rightsubnet's - ikev1
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
VR
StrongSwan
~~~

##### CLOUDSTACK VERSION

~~~
ACS 4.11.2.0
~~~

##### CONFIGURATION

Advanced Network
VPC

##### SUMMARY

After upgrading ACS from 4.9.3 (openswan) to 4.11.2 (strongswan), all VPNs with multiple networks have stopped working. Only one of the networks declared in the encryption domain passed traffic.

rightsubnet=192.168.198.0/23,192.168.208.0/23,192.168.170.0/23,192.168.234.0/23,192.168.69.0/24

I changed the configuration manually by creating different Child SAs, one for each network, now all networks work.
https://lists.strongswan.org/pipermail/users/2015-November/008966.html

Example:
#conn for vpn-4.3.2.1
conn vpn-4.3.2.1
 left=1.2.3.4
 leftsubnet=192.168.101.0/24
 right=4.3.2.1
 type=tunnel 
 authby=secret 
 keyexchange=ike
 ike=aes128-sha1-modp1024
 ikelifetime=1h 
 esp=aes128-sha1-modp1024
 lifetime=8h 
 keyingtries=2
 auto=start
 forceencaps=no
 dpddelay=30
 dpdtimeout=120
 dpdaction=restart

conn net-192.168.198.0
 also=vpn-4.3.2.1
 rightsubnet=192.168.198.0/23
 auto=start

conn net-192.168.208.0
 also=vpn-4.3.2.1
 rightsubnet=192.168.208.0/23
 auto=start

conn net-192.168.170.0
 also=vpn-4.3.2.1
 rightsubnet=192.168.170.0/23
 auto=start

conn net-192.168.234.0
 also=vpn-4.3.2.1
 rightsubnet=192.168.234.0/23
 auto=start

conn net-192.168.69.0
 also=vpn-4.3.2.1
 rightsubnet=192.168.69.0/24
 auto=start

### PR DESCRIPTION
**Pull request by DaanHoogland**
IKE version allows selecting ike (autoselect), ikev1, or ikev2.
Split connections gives an option of separating the first right subnet from the rest, and kicking out individual statements for each right subnet for better cross-compatibility.

update per PR suggestion

Fixes #3138

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
cc @ggoodrich-ipp

**Comment by rhtyd** [comment]
@blueorangutan ui

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by nvazquez** [comment]
@DaanHoogland can you please resolve the conflicts?

**Comment by rhtyd** [comment]
@blueorangutan package

**Review by rhtyd — APPROVED** [review]
LGTM - didn't test - this needs manual testing of ikev1+2 based s2s vpn cc @Pearl1594

File: api/src/main/java/org/apache/cloudstack/api/command/user/vpn/UpdateVpnCustomerGatewayCmd.java [hunk 1]
```diff
@@ -140,6 +146,14 @@ public Boolean getDpd() {
 
     public Boolean getEncap() { return encap; }
 
+    public boolean getSplitConnections() {
```
Comment (Pearl1594) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vpn/UpdateVpnCustomerGatewayCmd.java:149]: Update of VPN gateway results in an NPE if splitconnection isn't passed, as the return type is `boolean` (primitive). We can either change this to `Boolean` or do the null check here.

```
2021-04-22 04:23:54,483 DEBUG [c.c.a.ApiServlet] (qtp1233705144-17:ctx-f384e62b ctx-1b68af67) (logid:5c3e9c53) ===END===  172.16.250.2 -- GET  id=10c9ef13-e7ab-4c40-8c8c-becd750f990a&name=vt2-right-gw1&gateway=10.0.52.84&cidrlist=10.2.0.0%2F24&ipsecpsk=password&ikepolicy=aes128-sha1%3Bmodp1536&ikelifetime=86400&ikeversion=ike&esppolicy=aes128-sha1&esplifetime=3600&command=updateVpnCustomerGateway&response=json
2021-04-22 04:23:54,492 ERROR [c.c.a.ApiAsyncJobDispatcher] (API-Job-Executor-20:ctx-39e2f033 job-78) (logid:f3cdea9e) Unexpected exception while executing org.apache.cloudstack.api.command.user.vpn.UpdateVpnCustomerGatewayCmd
java.lang.NullPointerException
	at org.apache.cloudstack.api.command.user.vpn.UpdateVpnCustomerGatewayCmd.getSplitConnections(UpdateVpnCustomerGatewayCmd.java:150)
	at com.cloud.network.vpn.Site2SiteVpnManagerImpl.updateCustomerGateway(Site2SiteVpnManagerImpl.java:482)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
```

Comment (DaanHoogland) [reply api/src/main/java/org/apache/cloudstack/api/command/user/vpn/UpdateVpnCustomerGatewayCmd.java:149]: I am changing this, I noticed one other occasion (for which the fiels is a basic type , so it won't give the NPE, I'd rather keep the interface the original author decided on so I'll do the null check in the method.

**Comment by kricud** [comment]
Did testing and found following things.

Problems:

Add VPN Customer Gateway > IKE Version > set:ikev2 > press:ok
GW is created and IKE Version is set to: ike
Workaround: edit VPN Customer Gateway > IKE Version > type:ikev2

Add VPN Customer Gateway > IKE Version > set:ikev1 > press:ok
GW is created and IKE Version is set to: ike
Workaround: edit VPN Customer Gateway > IKE Version > type:ikev1

Add VPN Customer Gateway > Edit VPN Customer Gateway > CIDR list > Command failed due to Internal Server Error
Workaround: Delete VPN Customer Gateway and create again with correct settings.

Suggestions:

Edit help text for Split Connections: Only for IKEv2, whether to split multiple right subnet CIDRs into multiple connection statements.
Source:https://wiki.strongswan.org/projects/strongswan/wiki/FAQ#Multiple-subnets-per-SA
Essence: If you use IKEv1, you need to be a roadwarrior and use the UNITY extension.

Add help text for IKE Version : Connections marked with ike will use IKEv2 when initiating, but accept any protocol version when responding.
Source:https://wiki.strongswan.org/projects/strongswan/wiki/connsection

**Comment by rhtyd** [comment]
Thanks for sharing test results @kricud

File: ui/src/views/network/CreateVpnCustomerGateway.vue [hunk 2]
```diff
@@ -317,7 +350,9 @@ export default {
           dpd: values.dpd,
           forceencap: values.forceencap,
           ikepolicy: ikepolicy,
-          esppolicy: esppolicy
+          esppolicy: esppolicy,
+          splitconnections: values.splitconnections,
+          ikeversion: values.ikeversion
```
Comment (Pearl1594) [inline ui/src/views/network/CreateVpnCustomerGateway.vue:355]:
```suggestion
          ikeversion: values.ikeVersions
```

This causes ikeversion to be null at the time of creation of the customer gateway - as it isn't passed, hence defaults to ike
image

**Review by Pearl1594 — CHANGES_REQUESTED** [review]

**Comment by rhtyd** [comment]
Closing, as the PR has moved to - https://github.com/apache/cloudstack/pull/4953 (Pearl is not committer, unable to push changes on this PR).

**Pull request closed without merging**